## Segmenting and clustering neighborhoods in Toronto



#### 1. Creating neighbourhood dataframe

In [4]:
import pandas as pd
import numpy as np

In [5]:
# importing table from wikipedia with pandas

df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

In [12]:
# removing Not assigned boroughs

df_dropped = df.drop(df[df['Borough'] == 'Not assigned'].index)
df_dropped.reset_index(drop=True)

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [13]:
#validating if no duplicated Postal codes entries exist

df_dropped['Postal Code'].value_counts()

M1E    1
M5A    1
M9A    1
M5S    1
M4H    1
      ..
M4J    1
M3L    1
M1L    1
M5R    1
M9R    1
Name: Postal Code, Length: 103, dtype: int64

No duplicated Postal codes exist! (103 postal codes all with count 1)

In [14]:
#validate if any Neighbourhood is Not assigned

df_dropped[df_dropped['Neighbourhood'] == 'Not assigned']



Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

**Dataframe is validated, assigning new name and validating shape**

In [15]:
toronto_hoods=df_dropped
toronto_hoods.shape

(103, 3)

## 2. Adding Latitude and Longitude to my DataFrame

In [10]:
from geopy.geocoders import Nominatim


In [11]:
# testing Nominatim

address='Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

address='M5G, Toronto, ON'

location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of M5G are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


AttributeError: 'NoneType' object has no attribute 'latitude'

Nominatim works with borough name but not with postal code, will try provided example

In [21]:
import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

postal_code='M5G'

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]

ModuleNotFoundError: No module named 'geocoder'

Getting error, will use csv instead

In [16]:
#donwloaded via browser, don't have wget installed

postal_coords=pd.read_csv('Geospatial_Coordinates.csv')
postal_coords

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [17]:
toronto_hoods=pd.merge(toronto_hoods,postal_coords,on='Postal Code')
toronto_hoods

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

**Ex 2 completed!**

## 3. Exploring and clustering neihborhoods in Toronto

In [18]:
import folium

Plotting Map of toronto with Postal Code locations:



In [19]:
# 1st getting Toronto coords

address='Toronto, ON'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# drawing map with code from lab

map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, label in zip(toronto_hoods['Latitude'], toronto_hoods['Longitude'], toronto_hoods['Postal Code']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [20]:
#Foursquare API stuff

CLIENT_ID = 'GTXNQLAPZJL4DTT2FLYPJCX2OM3XYRJJ4CLXMQRUFCIHFJ2K' # your Foursquare ID
CLIENT_SECRET = 'J3RSGNVPLVXDPIKCLRYA1UMN3K2YPX5B5N2V3VSOTRB0HQ0O' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value


In [21]:
# Experimenting for 1st Postcal Code in df

pc_latitude=toronto_hoods.loc[0,'Latitude']
pc_longitude=toronto_hoods.loc[0,'Longitude']

# test list seemed short so importing 750 isntead of 500m
radius=750

print('lat : {} , lon : {}'.format(pc_latitude,pc_longitude))

#url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, pc_latitude, pc_longitude, VERSION, radius, LIMIT)
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, pc_latitude, pc_longitude, VERSION, radius, LIMIT)
url


lat : 43.7532586 , lon : -79.3296565


'https://api.foursquare.com/v2/venues/explore?client_id=GTXNQLAPZJL4DTT2FLYPJCX2OM3XYRJJ4CLXMQRUFCIHFJ2K&client_secret=J3RSGNVPLVXDPIKCLRYA1UMN3K2YPX5B5N2V3VSOTRB0HQ0O&ll=43.7532586,-79.3296565&v=20180605&radius=750&limit=100'

In [22]:
import requests

results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f9aa48ca106903c94db9cb9'},
 'response': {'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 4,
  'suggestedBounds': {'ne': {'lat': 43.760008606750006,
    'lng': -79.32032910977391},
   'sw': {'lat': 43.74650859325, 'lng': -79.33898389022609}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.751976046055574,
          'lng': -79.33214044722958}],
        'distance': 245,
        'cc': 'CA

In [23]:
# function that extracts the category of the venue; copied from last lab
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']
    
venues = results['response']['groups'][0]['items']
    
nearby_venues = pd.json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0    Brookbanks Park               Park  43.751976 -79.332140
1      Variety Store  Food & Drink Shop  43.751974 -79.333114
2  DVP at York Mills       Intersection  43.758899 -79.334099
3    TTC Stop #09083           Bus Stop  43.759655 -79.332223

In [44]:
nearby_venues.shape[0]

4

In [24]:
#function to get all nearby_venues for all postal codes

def getNearbyVenues(postal_code, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for postal_code, lat, lng in zip(postal_code, latitudes, longitudes):
        print(postal_code)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            postal_code, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code', 
                  'PC Latitude', 
                  'PC Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [25]:
toronto_venues = getNearbyVenues(postal_code=toronto_hoods['Postal Code'],
                                latitudes=toronto_hoods['Latitude'],
                                longitudes=toronto_hoods['Longitude'])

M3A
M4A
M5A
M6A
M7A
M9A
M1B
M3B
M4B
M5B
M6B
M9B
M1C
M3C
M4C
M5C
M6C
M9C
M1E
M4E
M5E
M6E
M1G
M4G
M5G
M6G
M1H
M2H
M3H
M4H
M5H
M6H
M1J
M2J
M3J
M4J
M5J
M6J
M1K
M2K
M3K
M4K
M5K
M6K
M1L
M2L
M3L
M4L
M5L
M6L
M9L
M1M
M2M
M3M
M4M
M5M
M6M
M9M
M1N
M2N
M3N
M4N
M5N
M6N
M9N
M1P
M2P
M4P
M5P
M6P
M9P
M1R
M2R
M4R
M5R
M6R
M7R
M9R
M1S
M4S
M5S
M6S
M1T
M4T
M5T
M1V
M4V
M5V
M8V
M9V
M1W
M4W
M5W
M8W
M9W
M1X
M4X
M5X
M8X
M4Y
M7Y
M8Y
M8Z


In [26]:
toronto_venues.shape

(2136, 7)

In [30]:
toronto_venues.groupby('Postal Code').count()

PC Latitude  PC Longitude  Venue  Venue Latitude  \
Postal Code                                                     
M1B                    1             1      1               1   
M1C                    2             2      2               2   
M1E                    8             8      8               8   
M1G                    4             4      4               4   
M1H                    8             8      8               8   
...                  ...           ...    ...             ...   
M9N                    2             2      2               2   
M9P                    6             6      6               6   
M9R                    4             4      4               4   
M9V                   10            10     10              10   
M9W                    4             4      4               4   

             Venue Longitude  Venue Category  
Postal Code                                   
M1B                        1               1  
M1C                        2               2  
M1E                        8               8  
M1G                        4               4  
M1H                        8               8  
...                      ...             ...  
M9N                        2               2  
M9P                        6               6  
M9R                        4               4  
M9V                       10              10  
M9W                        4               4  

[100 rows x 6 columns]

In [32]:
len(toronto_venues['Venue Category'].unique())

273

In [35]:
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Postal Code'] = toronto_venues['Postal Code'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_grouped= toronto_onehot.groupby('Postal Code').mean().reset_index()
toronto_grouped

Postal Code  Accessories Store  Afghan Restaurant  Airport  \
0          M1B                0.0                0.0      0.0   
1          M1C                0.0                0.0      0.0   
2          M1E                0.0                0.0      0.0   
3          M1G                0.0                0.0      0.0   
4          M1H                0.0                0.0      0.0   
..         ...                ...                ...      ...   
95         M9N                0.0                0.0      0.0   
96         M9P                0.0                0.0      0.0   
97         M9R                0.0                0.0      0.0   
98         M9V                0.0                0.0      0.0   
99         M9W                0.0                0.0      0.0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                  0.0           0.0             0.0              0.0   
1                  0.0           0.0             0.0              0.0   
2                  0.0           0.0             0.0              0.0   
3                  0.0           0.0             0.0              0.0   
4                  0.0           0.0             0.0              0.0   
..                 ...           ...             ...              ...   
95                 0.0           0.0             0.0              0.0   
96                 0.0           0.0             0.0              0.0   
97                 0.0           0.0             0.0              0.0   
98                 0.0           0.0             0.0              0.0   
99                 0.0           0.0             0.0              0.0   

    Airport Terminal  American Restaurant  ...  Train Station  \
0                0.0                  0.0  ...            0.0   
1                0.0                  0.0  ...            0.0   
2                0.0                  0.0  ...            0.0   
3                0.0                  0.0  ...            0.0   
4                0.0                  0.0  ...            0.0   
..               ...                  ...  ...            ...   
95               0.0                  0.0  ...            0.0   
96               0.0                  0.0  ...            0.0   
97               0.0                  0.0  ...            0.0   
98               0.0                  0.0  ...            0.0   
99               0.0                  0.0  ...            0.0   

    Turkish Restaurant  Vegetarian / Vegan Restaurant  Video Game Store  \
0                  0.0                            0.0               0.0   
1                  0.0                            0.0               0.0   
2                  0.0                            0.0               0.0   
3                  0.0                            0.0               0.0   
4                  0.0                            0.0               0.0   
..                 ...                            ...               ...   
95                 0.0                            0.0               0.0   
96                 0.0                            0.0               0.0   
97                 0.0                            0.0               0.0   
98                 0.0                            0.0               0.0   
99                 0.0                            0.0               0.0   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                     0.0              0.0       0.0          0.0   
1                     0.0              0.0       0.0          0.0   
2                     0.0              0.0       0.0          0.0   
3                     0.0              0.0       0.0          0.0   
4                     0.0              0.0       0.0          0.0   
..                    ...              ...       ...          ...   
95                    0.0              0.0       0.0          0.0   
96                    0.0              0.0       0.0          0.0   
97                    0.0              0.0

In [36]:
#copying from last lab

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postal Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
pc_venues_sorted = pd.DataFrame(columns=columns)
pc_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']

for ind in np.arange(toronto_grouped.shape[0]):
    pc_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

pc_venues_sorted.head()

Postal Code       1st Most Common Venue 2nd Most Common Venue  \
0         M1B        Fast Food Restaurant   Dumpling Restaurant   
1         M1C  Construction & Landscaping                   Bar   
2         M1E              Breakfast Spot            Restaurant   
3         M1G                 Coffee Shop    Mexican Restaurant   
4         M1H                 Gas Station   Fried Chicken Joint   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
0        Discount Store          Distribution Center               Dog Run   
1           Yoga Studio  Eastern European Restaurant               Dog Run   
2     Electronics Store               Medical Center   Rental Car Location   
3     Korean Restaurant          Distribution Center               Dog Run   
4                Bakery                         Bank    Athletics & Sports   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant            Donut Shop             Drugstore   
1      Doner Restaurant            Donut Shop             Drugstore   
2          Intersection    Mexican Restaurant                  Bank   
3      Doner Restaurant            Donut Shop             Drugstore   
4       Thai Restaurant  Caribbean Restaurant      Hakka Restaurant   

         9th Most Common Venue       10th Most Common Venue  
0  Eastern European Restaurant      Health & Beauty Service  
1          Dumpling Restaurant            Electronics Store  
2                  Yoga Studio             Doner Restaurant  
3          Dumpling Restaurant                  Yoga Studio  
4            Electronics Store  Eastern European Restaurant

In [58]:
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

n_clusters = 5

abc=toronto_grouped.drop('Postal Code', 1)
abc

kmeans = KMeans(n_clusters=n_clusters, random_state=2).fit(abc)

kmeans.label_


array([0, 1, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 2, 0, 3, 3, 3, 4, 3, 2,
       0, 2, 3, 3, 3, 3, 2, 3, 3, 0, 3, 0, 3, 3, 3, 3, 2, 3, 3, 3, 2, 3,
       3, 3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 0, 3, 2, 3, 3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0,
       3, 1, 3, 3, 3, 0, 1, 2, 3, 0, 0, 3])

In [86]:
tor_merged = toronto_hoods
pc_venues_sorted.drop('Cluster Labels', axis=1, inplace=True)
pc_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)


pc_venues_sorted
tor_merged = tor_merged.join(pc_venues_sorted.set_index('Postal Code'), on='Postal Code')

#Postal Codes with no venues got assigned cluster nan so changing those to cluster 5

tor_merged['Cluster Labels'].fillna(5, inplace=True)
tor_merged = tor_merged.astype({'Cluster Labels': int})
tor_merged['Cluster Labels'].unique()




array([2, 3, 5, 0, 1, 4])

Now we have no nulls and clusters are int, plotting next

In [87]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(n_clusters)
ys = [i + x + (i*x)**2 for i in range(n_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Postal Code'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

And we're done!